File: 17. Diameter of tree.py
class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def diameter_util(root):
    if not root: return (0,0) # height, diameter
    lh, ld = diameter_util(root.left)
    rh, rd = diameter_util(root.right)
    height = 1 + max(lh, rh)
    diam = max(lh+rh+1, max(ld, rd))
    return (height, diam)

def diameter(root):
    return diameter_util(root)[1]

if __name__=='__main__':
    root=Node(1); root.left=Node(2); root.right=Node(3); root.left.left=Node(4); root.left.right=Node(5)
    print('Diameter:', diameter(root))

File: 17. Diameter of tree.cpp
#/*
 Task: Diameter of tree
 Time Complexity: O(n)
 Space Complexity: O(h) — recursion stack
*/
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
int diameterUtil(Node* root, int& res)
{
    if (!root)
        return 0;
    int l = diameterUtil(root->left, res);
    int r = diameterUtil(root->right, res);
    res = max(res, l + r + 1);
    return 1 + max(l, r);
}
int diameter(Node* root)
{
    int res = 0;
    diameterUtil(root, res);
    return res;
}
int main()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    cout << "Diameter: " << diameter(root) << endl;
    return 0;
}
// sc tc
